ports/stm/common-hal/canio/Listener.c

```diff
@@ -0,0 +1,315 @@
+}
+
+STATIC void install_extended_filter(canio_listener_obj_t *self, canio_match_obj_t *match) {
+    int bank = next_filter(self->can);
+
+    // filter is already deactivated, so we skip deactivating it here
+    // CLEAR_BIT(self->can->filter_hw->FA1R, bank);
+
+    self->can->filter_hw->sFilterRegister[bank].FR1 =
+        ((match->id << 3) | FILTER32_IDE);
+    self->can->filter_hw->sFilterRegister[bank].FR2 =
+        ((match->mask << 3) | FILTER32_IDE);
+
+    // filter mode: 0 = mask
+    // (this bit should be clear already, we never set it; but just in case)
+    CLEAR_BIT(self->can->filter_hw->FM1R, 1 << bank);
+    // filter scale: 1 = 32 bits
+    SET_BIT(self->can->filter_hw->FS1R, 1 << bank);
+    // fifo assignment: 1 = FIFO 1
+    if (self->fifo_idx) {
+        SET_BIT(self->can->filter_hw->FFA1R, 1 << bank);
+    } else {
+        CLEAR_BIT(self->can->filter_hw->FFA1R, 1 << bank);
+    }
+
+    // filter activation: 1 = enabled
+    SET_BIT(self->can->filter_hw->FA1R, 1 << bank);
+}
+
+STATIC void install_all_match_filter(canio_listener_obj_t *self) {
+    int bank = next_filter(self->can);
+
+    // filter is already deactivated, so we skip deactivating it here
+    // CLEAR_BIT(self->can->filter_hw->FA1R, bank);
+
+    self->can->filter_hw->sFilterRegister[bank].FR1 = 0;
+    self->can->filter_hw->sFilterRegister[bank].FR2 = 0;
+
+    // filter mode: 0 = mask
+    // (this bit should be clear already, we never set it; but just in case)
+    CLEAR_BIT(self->can->filter_hw->FM1R, bank);
+    // filter scale: 1 = 32 bits
+    SET_BIT(self->can->filter_hw->FS1R, bank);
+    // fifo assignment: 1 = FIFO 1
+    if (self->fifo_idx) {
+        SET_BIT(self->can->filter_hw->FFA1R, bank);
+    } else {
+        CLEAR_BIT(self->can->filter_hw->FFA1R, bank);
+    }
+
+    // filter activation: 1 = enabled
+    SET_BIT(self->can->filter_hw->FA1R, (1 << bank));
+}
+
+
+#define NO_ADDRESS (-1)
+void set_filters(canio_listener_obj_t *self, size_t nmatch, canio_match_obj_t **matches) {
+    allow_filter_change(self->can);
+    if (!nmatch) {
+        install_all_match_filter(self);
+    } else {
+        canio_match_obj_t *first_match = NULL;
+        for(size_t i = 0; i<nmatch; i++) {
+            if(matches[i]->extended) {
+                install_extended_filter(self, matches[i]);
+            } else {
+                if (first_match) {
+                    install_standard_filter(self, first_match, matches[i]);
+                    first_match = NULL;
+                } else {
+                    first_match = matches[i];
+                }
+            }
+        }
+        if (first_match) {
+            install_standard_filter(self, first_match, first_match);
+        }
+    }
+    prevent_filter_change(self->can);
+}
+
+
+void common_hal_canio_listener_construct(canio_listener_obj_t *self, canio_can_obj_t *can, size_t nmatch, canio_match_obj_t **matches, float timeout) {
+    if (!can->fifo0_in_use) {
```

Comment:
Is the expectation that a canio stm32f405 user will always need to create 2 listeners?

Reply:
I'm not sure if this is what you're asking, but I believe the bus needs two devices to function as n active mode <https://github.com/adafruit/Adafruit_CircuitPython_MCP2515/issues/21#issuecomment-1615381996>

Reply:
In this repo, if I understand this correctly, a separate listener object is required to use FIFO1. I was comparing this implementation with another [library ](https://github.com/stm32-rs/bxcan/blob/master/src/lib.rs#L691)(Rust), and noticed that in their receive handler, they check both FIFOs in one check.

Reply:
It is not clear to me which listener will get which messages when there is no matching/filtering setup. I suppose the likely recommendation is to use message id filtering.

Aka, incorrect usage (that will lead to weird behavior):
```
listenerFIFO0 = can.listen(timeout=.1)
listenerFIFO1 = can.listen(timeout=.1)
```

Correct usage:
```
listenerFIFO0 = can.listen(matches=[<list of message ids>], timeout=.1)
listenerFIFO1 = can.listen(matches=[<list of message ids different than FIFO0>], timeout=.1)
```

I have not tested this though.